### DISCUSSION TITLE
Need help for KafkaSteps.autoSeekTopics

### DISCUSSION THREAD

**Discussion by chombymat** [Decathlon/tzatziki#326] [Q&A] [answered]
Hello everyone,

When my scenarios follow one another, I cannot correctly check the content of the topic in which I produce.
Unitary the tests pass.

I use the following step:
Then the topicName topic contains 0 RecordName

I think I have a configuration problem.
I have specified in the Initializer KafkaSteps.autoSeekTopics("topicName");

Can you help me?

If you need more informations please feel free to ask me.

Thanks a lot.

**Comment by alexandrepa** [comment]
Hello, can you be more precise on the exception you have. There is more than 0 messages in the topic or he can't read the topic at all ?

**Reply by chombymat** [reply]
There is more than 0 messages in the topic even if the scenario contains only the assertion step like this one:
Scenario: we can assert that no message has been sent to a topic
    * the exposed-users topic contains 0 user

My test found 2 messages

**Reply by alexandrepa** [reply]
You possibly have a background task (thread) that are still pushing messages after the end of the 1st test (and the autoseek) and before the 2nd test ?

**Reply by chombymat** [reply]
I don't think so, in my case the scenario that has an error found 2 messages because 2 tests are executed before it, if I execute 1 test before it finds 1 message.
If the messages are still pushed after, I think the number of messages found in my test scenario should be higher right ?

**Reply by alexandrepa** [reply]
Are you creating a new kafka consumer for each test ? Or with a new group-id ? Because autoseek works only with shared kafka consumer between tests.

**Reply by chombymat** [reply]
I'm not sure but i think not, the clientId and groupId are the same in all scenario. I don't feel like I created a specific configuration for each test.
Do you have any verification that are more valid than the logs ?

**Reply by alexandrepa** [reply]
You can add this in the background :
* a com.decathlon.tzatziki.steps.KafkaSteps logger set to DEBUG

and see if you have some log about topic autoseek :
setting offset of %s topic to %s

**Reply by chombymat** [reply]
I had test with:
* a root logger set to DEBUG
And another with:
* a com.decathlon.tzatziki.steps.KafkaSteps logger set to DEBUG

i don't have logs that match exactly "setting offset of %s topic to %s"

I have some logs that I hope might be relevant
They are listed in order of appearance:
1. [Consumer clientId=test-topicConsumed-0, groupId=test-group-topicConsumed] Resetting offset for partition topicConsumed-0 to position FetchPosition{offset=0, offsetEpoch=Optional.empty, currentLeader=LeaderAndEpoch{leader=Optional[localhost:60204 (id: 0 rack: null)], epoch=0}}.
2. [org.springframework.kafka.KafkaListenerEndpointContainer#0-0-C-1] INFO  o.a.k.c.c.i.SubscriptionState - [Consumer clientId=test-topicConsumed-0, groupId=test-group-topicConsumed] Resetting offset for partition topicConsumed-0 to position FetchPosition{offset=0, offsetEpoch=Optional.empty, currentLeader=LeaderAndEpoch{leader=Optional[localhost:60365 (id: 0 rack: null)], epoch=0}}.
3. [INFO ] [awaitility-thread] o.a.k.c.c.i.SubscriptionState - [Consumer clientId=test, groupId=d831b7d9-7246-4a55-9f59-7cbcdcfe3850_avro_topicPublicated] Seeking to earliest offset of partition topicPublicated-0
4. [INFO ] [awaitility-thread] o.a.k.c.c.i.SubscriptionState - [Consumer clientId=test, groupId=d831b7d9-7246-4a55-9f59-7cbcdcfe3850_avro_topicPublicated] Resetting offset for partition topicPublicated-0 to position FetchPosition{offset=0, offsetEpoch=Optional.empty, currentLeader=LeaderAndEpoch{leader=Optional[localhost:60365 (id: 0 rack: null)], epoch=0}}.
5. [INFO ] [awaitility-thread] o.a.k.c.c.i.LegacyKafkaConsumer - [Consumer clientId=test, groupId=d831b7d9-7246-4a55-9f59-7cbcdcfe3850_avro_topicPublicated] Seeking to offset 1 for partition topicPublicated-0
6. [INFO ] [main] o.a.k.c.c.i.LegacyKafkaConsumer - [Consumer clientId=test, groupId=d831b7d9-7246-4a55-9f59-7cbcdcfe3850_avro_topicPublicated] Seeking to offset 0 for partition topicPublicated-0

I have noticied that i can use the step 
When this event is published on the topicPublicated topic
but not 
When this event is consumed from the topicPublicated topic
cause by: org.awaitility.core.ConditionTimeoutException: Condition com.decathlon.tzatziki.kafka.KafkaInterceptor$$Lambda/0x000001c4e32d34e0 was not fulfilled within 10 seconds.

I don't know if it's relevant with the subject. If not, don't try to help me on this, i'll figure it out later.

**Reply by alexandrepa** [reply]
I maybe have an idea of the issue, the `@Aspect` annotation conf are not taken in account.
Can you add this annotation : 
`@ComponentScan(basePackages = "com.decathlon")`

On a `@Configuration `class in your test folder

**Reply by chombymat** [reply]
I have added the @ComponentScan and nothing changed.

I share my class, maybe i missed something:
@ActiveProfiles("test")
@RunWith(Cucumber.class)
@CucumberContextConfiguration
@ComponentScan(basePackages = "com.decathlon")
@CucumberOptions(
        plugin = {"pretty", "json:target/cucumber.json"},
        extraGlue = {"com.decathlon.tzatziki.steps"},
        tags = "not @ignore")
@SpringBootTest(
        classes = Application.class,
        webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@ContextConfiguration(initializers = ApplicationTest.InitializerTest.class)
public class ApplicationTest {

    static final MongoDBContainer mongoDBContainer =
            new MongoDBContainer(DockerImageName.parse("mongo:4.2-bionic"));

    static class InitializerTest
            implements ApplicationContextInitializer<ConfigurableApplicationContext> {

        public void initialize(ConfigurableApplicationContext configurableApplicationContext) {
            KafkaSteps.start();

            mongoDBContainer.start();
            TestPropertyValues.of(
                            "spring.data.mongodb.uri=" + mongoDBContainer.getReplicaSetUrl(),
                            "spring.kafka.consumer.auto-offset-reset=earliest",
                            "spring.kafka.bootstrap-servers=" + KafkaSteps.bootstrapServers(),
                            "spring.kafka.properties.schema.registry.url="
                                    + KafkaSteps.schemaRegistryUrl())
                    .applyTo(configurableApplicationContext.getEnvironment());
            KafkaSteps.autoSeekTopics("topicPublicated");
        }
    }
}

**Reply by alexandrepa** [reply]
The `@ComponentScan(basePackages = "com.decathlon")` annotation must be set on a class that is annotated with `@Configuration`, can you add it on one of your conf class somewhere in your app to see if it's the problem ?

**Reply by alexandrepa** [reply, marked as answer]
Another way to resolve this that should work, is to add `classes = KafkaInterceptor.class` to the `@ContextConfiguration` annotation.

Should be some thing like this in your code :

`@ContextConfiguration(initializers = ApplicationTest.InitializerTest.class, classes = KafkaInterceptor.class)`

**Reply by chombymat** [reply]
Yes, both of your suggestions work !
That's fix the autoSeek problème AND the published on/consumed from !
Thank you very much for your help and time, I don't think I would have figured this out on my own !

**Reply by alexandrepa** [reply]
No problem, thank you to raise this issue, I'm going to update the README to be clear about that if the app is not in the com.decathlon package.
